server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -58,58 +65,76 @@ public static MailDto fromMail(Mail mail, List<AdditionalFields> additionalField
             fetchMessageSize(additionalFields, mail));
     }
 
-    private static Optional<Long> fetchMessageSize(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
-        if (!additionalFields.contains(AdditionalFields.MESSAGE_SIZE)) {
+    private static Optional<Long> fetchMessageSize(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.MESSAGE_SIZE)) {
             return Optional.empty();
         }
         try {
-            return Optional.ofNullable(new Long(mail.getMessageSize()));
+            return Optional.of(new Long(mail.getMessageSize()));
         } catch (MessagingException e) {
-            throw new MissingRequestedField("messageSize");
+            throw new InnaccessibleFieldException(AdditionalField.MESSAGE_SIZE);
         }
     }
 
-    private static Optional<String> fetchBody(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
-        if (!additionalFields.contains(AdditionalFields.BODY)) {
+    private static Optional<String> fetchBody(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.BODY)) {
             return Optional.empty();
         }
+
         try {
-            return Optional.ofNullable(mail.getMessage().getContent().toString());
-        } catch (IOException | MessagingException | NullPointerException e) {
-            throw new MissingRequestedField("body");
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function((Message message) -> extractor.extract(message)).sneakyThrow())
+                    .flatMap(extractBody());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.BODY);
         }
     }
 
-    private static Optional<Map<String, List<String>>> fetchHeaders(List<AdditionalFields> additionalFields, Mail mail) throws MissingRequestedField {
-        if (!additionalFields.contains(AdditionalFields.HEADERS)) {
+    private static Function<? super MessageContent, Optional<String>> extractBody() {
```

Comment:
IMO we should have 2 additional fields : instead of `body`, we should have `textBody` and `htmlBody`